File: src/Python/array/84.柱状图中最大的矩形.py
#
# @lc app=leetcode.cn id=84 lang=python
#
# [84] 柱状图中最大的矩形
#

# @lc code=start
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        stack = []
        heights = [0] + heights + [0]
        res = 0
        for i in range(len(heights)):
            while stack and heights[stack[-1]] > heights[i]:
                left_index = stack.pop()
                # heights[i] * (right_i - left_i -1)
                res = max(res, (i - stack[-1] - 1) * heights[left_index])
            stack.append(i)
        return res
# @lc code=end

File: src/Java/stack/84.柱状图中最大的矩形.java
/*
 * @lc app=leetcode.cn id=84 lang=java
 *
 * [84] 柱状图中最大的矩形
 */

// @lc code=start
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<Integer>();
        int[] newHeights = new int[heights.length + 2];
        int res = 0;
        for (int i = 1; i < heights.length + 1; i++) newHeights[i] = heights[i - 1];
        for (int i = 0; i < newHeights.length; i++) {
            while (!stack.isEmpty() && newHeights[stack.peek()] > newHeights[i]) {
                int cur = stack.pop();
                res = Math.max(res, (i - stack.peek() - 1) * newHeights[cur]);
            }
            stack.push(i);
        }
        return res;
    }
}
// @lc code=end

